pyllamacpp/model.py

```diff
@@ -158,28 +168,41 @@ def generate(self,
             self._last_n_tokens.append(tok)
 
         predicted_word = ""
-        n_remain = 0
-
         tokens_queue = []
+        self.n_consumed_tokens = 0
 
         while infinite_generation or predicted_token != pp.llama_token_eos():
             if len(predicted_tokens) > 0:
                 if (pp.llama_eval(self._ctx,
                                   predicted_tokens,
                                   len(predicted_tokens),
-                                  self._n_past,
+                                  self.n_consumed_tokens,
                                   n_threads)):
                     raise Exception("failed to eval the model!")
-                self._n_past += len(predicted_tokens)
+                self.n_consumed_tokens += len(predicted_tokens)
                 predicted_tokens.clear()
 
+            input_prompt = self._prompt_context_tokens + self._last_n_tokens[self._n_ctx - self.n_consumed_tokens:]
+            """
+            # For debug
+            print('--------------------- prompt ----------------------')
+            prmpt = pp.llama_tokens_to_str(self._ctx, input_prompt)
+            print(prmpt)
+            print()
+            print('--------------------- pred ----------------------')
+            print(pp.llama_tokens_to_str(self._ctx, predicted_tokens))
+            print()
+
+            """
+            # The context should stay fixed and never vanish
```

Comment:
It is already there (with `n_past`)!
Have you read the `generate` function in the cpp side, as originally implemented in `llama.cpp`?